exercises/palindrome-products/test/Tests.hs

```diff
@@ -13,21 +13,36 @@ main = hspecWith defaultConfig {configFastFail = True} specs
 specs :: Spec
 specs = for_ cases test
   where
-    test (desc, minFactor, maxFactor, sPal, sPalFactors, lPal, lPalFactors) =
+    test (desc, minFactor, maxFactor, Just (sPal, sPalFactors, lPal, lPalFactors)) =
       describe desc $ do
         let sortPair (a, b)  = if a < b then (a, b) else (b, a)
         let normalize        = sort . nub . map sortPair
+        describe "smallestPalindrome" $
+          case smallestPalindrome minFactor maxFactor of
+            Just (value, factors) -> do
+              it "value"   $ value             `shouldBe` sPal
+              it "factors" $ normalize factors `shouldBe` sPalFactors
+            Nothing ->
+              it "result"  $ Nothing           `shouldBe` Just (sPal, [sPalFactors])
```

Comment:
I've refactored `smallestPalindrome` and `largestPalindrome` testing into a single function.

The reason I've added these exclusive cases is that I wanted to desctructure the result when possible in all 4 cases. My Haskell isn't good enough to figure out how do this more efficiently than using explicit cases ☺.

If I understand your proposal correctly, I don't think this would work. The type of `fmap fst result` is `Just Int`, which we then need to destructure again, or confuse the student by suddenly having a `Just` appear in the test output.